activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb

```diff
@@ -128,28 +141,61 @@ def execute(sql, name = nil)
         super
       end
 
-      def exec_query(sql, name = 'SQL', binds = [], prepare: false)
-        result = execute(sql, name)
-        @connection.next_result while @connection.more_results?
-        ActiveRecord::Result.new(result.fields, result.to_a)
-      end
-
-      def exec_insert(sql, name, binds, pk = nil, sequence_name = nil)
-        execute to_sql(sql, binds), name
-      end
-
-      def exec_delete(sql, name, binds)
-        execute to_sql(sql, binds), name
-        @connection.affected_rows
+      def exec_query(sql, name = 'SQL', binds = [], prepare: false) # :nodoc:
+        if without_prepared_statement?(binds)
+          execute_and_free(sql, name) do |result|
+            ActiveRecord::Result.new(result.fields, result.to_a) if result
+          end
+        else
+          exec_stmt_and_free(sql, name, binds, cache_stmt: prepare) do |_, result|
+            ActiveRecord::Result.new(result.fields, result.to_a) if result
```

Comment:
In mysql2 0.4.3, `result.fields` causes segv when num of rows is zero.
This issue was fixed on master by https://github.com/brianmario/mysql2/pull/741.
@sodabrew Could you release mysql2 0.4.4?